openhands/server/routes/git.py

```diff
@@ -244,67 +239,749 @@ async def get_repository_branches(
+    microagents_path: str,
+) -> list[MicroagentResponse]:
+    """Process .md files from GitLab tree items without fetching content."""
+    microagents = []
+
+    for item in tree_items:
+        if (
+            item['type'] == 'blob'
+            and item['name'].endswith('.md')
+            and item['name'] != 'README.md'
+        ):
+            try:
+                # Just create response from file metadata without fetching content
+                microagents.append(
+                    _create_microagent_response(
+                        item['name'], ProviderType.GITLAB, item['path']
+                    )
+                )
+            except Exception as e:
+                logger.warning(f'Error processing microagent {item["name"]}: {str(e)}')
+
+    return microagents
+
+
+async def _process_bitbucket_md_files(
+    client: httpx.AsyncClient,
+    directory_data: dict,
+    repository_name: str,
+    main_branch: str,
+    headers: dict[str, str],
+    microagents_path: str,
+) -> list[MicroagentResponse]:
+    """Process .md files from Bitbucket directory data without fetching content."""
+    microagents = []
+
+    if 'values' in directory_data:
+        for item in directory_data['values']:
+            if (
+                item['type'] == 'commit_file'
+                and item['path'].endswith('.md')
+                and not item['path'].endswith('README.md')
+            ):
+                try:
+                    # Extract file name from path
+                    file_name = item['path'].split('/')[-1]
+
+                    # Just create response from file metadata without fetching content
+                    microagents.append(
+                        _create_microagent_response(
+                            file_name, ProviderType.BITBUCKET, item['path']
+                        )
+                    )
+                except Exception as e:
+                    logger.warning(
+                        f'Error processing microagent {item["path"]}: {str(e)}'
+                    )
+
+    return microagents
+
+
+async def _fetch_github_file_content(
+    repository_name: str,
+    file_path: str,
+    provider_handler: ProviderHandler,
+) -> str:
+    """Fetch individual file content from GitHub repository.
+
+    Args:
+        repository_name: Repository name in format 'owner/repo'
+        file_path: Path to the file within the repository
+        provider_handler: Provider handler for authentication
+
+    Returns:
+        File content as string
+
+    Raises:
+        RuntimeError: If file cannot be fetched or doesn't exist
+    """
+    github_token = None
+    if (
+        provider_handler.provider_tokens
+        and ProviderType.GITHUB in provider_handler.provider_tokens
+    ):
+        token = provider_handler.provider_tokens[ProviderType.GITHUB].token
+        if token is not None:
+            github_token = token.get_secret_value()
+
+    headers = _get_github_headers(github_token)
```

Comment:
the provider handler handles this logic as well - see the `_get_service` method for the class

ideally we have one method per service class to ping specific api routes

then the provider handler will instantiate the classes and get the results from the service class methods